File: src/main/java/net/sf/rails/game/correct/CorrectionManager.java
package net.sf.rails.game.correct;

import java.util.ArrayList;
import java.util.List;

import net.sf.rails.common.DisplayBuffer;
import net.sf.rails.common.LocalText;
import net.sf.rails.common.ReportBuffer;
import net.sf.rails.game.GameManager;
import net.sf.rails.game.RailsAbstractItem;
import net.sf.rails.game.state.BooleanState;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


public abstract class CorrectionManager extends RailsAbstractItem {
    
    private final CorrectionType correctionType;
    private final BooleanState active = BooleanState.create(this, "active");
    
    protected static Logger log =
        LoggerFactory.getLogger(CorrectionManager.class);

    protected CorrectionManager(GameManager parent, CorrectionType ct) {
        super(parent, ct.name());
        correctionType = ct; 
    }
    
    @Override
    public GameManager getParent() {
        return (GameManager)super.getParent();
    }

    public CorrectionType getCorrectionType() {
        return correctionType;
    }

    public boolean isActive(){
        return active.value();
    }
    
    public List<CorrectionAction> createCorrections() {

        List<CorrectionAction> actions = new ArrayList<CorrectionAction>();
        actions.add(new CorrectionModeAction(getCorrectionType(), isActive()));

        return actions;
    }
   
    /** calls all executeAction */
    public boolean executeCorrection(CorrectionAction action){
        if (action instanceof CorrectionModeAction)  
            return execute((CorrectionModeAction) action);
        else {
            log.debug("This correction action is not registered.");
            return false;
        }
    }
    
    private boolean execute(CorrectionModeAction action) {
        
        
        if (!isActive()) {
            String text = LocalText.getText("CorrectionModeActivate",
                    getRoot().getPlayerManager().getCurrentPlayer().getId(),
                    LocalText.getText(getCorrectionType().name())
            );
            ReportBuffer.add(this, text);
            DisplayBuffer.add(this, text);
        }
        else {
            ReportBuffer.add(this, LocalText.getText("CorrectionModeDeactivate",
                    getRoot().getPlayerManager().getCurrentPlayer().getId(),
                    LocalText.getText(getCorrectionType().name())
            ));
        }
        active.set(!active.value());
     
        return true;
    }

    /* dummy to capture the non-supported actions */
    protected boolean execute(CorrectionAction action) {
        log.debug("The chosen action is not implemented in the registered manager");
        return false;
    }
    

    public boolean equals(CorrectionManager cm) {
        return (this.getParent() == cm.getParent() 
                && this.correctionType == cm.correctionType);
    }
}

File: dotnet/RailsLib.Net/Rails/Game/Correct/CorrectionManager.cs
using GameLib.Net.Common;
using GameLib.Net.Game;
using GameLib.Net.Game.State;
using System;
using System.Collections.Generic;
using System.Text;

namespace GameLib.Rails.Game.Correct
{
    abstract public class CorrectionManager : RailsAbstractItem, IEquatable<CorrectionManager>
    {
        private CorrectionType correctionType;
        private BooleanState active;


        protected static Logger<CorrectionManager> log = new Logger<CorrectionManager>();

    protected CorrectionManager(GameManager parent, CorrectionType ct) : base(parent, ct.Name)
        {
            active = BooleanState.Create(this, "active");
            correctionType = ct;
        }

        new public GameManager Parent
        {
            get
            {
                return (GameManager)base.Parent;
            }
        }

        public CorrectionType CorrectionType
        {
            get
            {
                return correctionType;
            }
        }

        public bool IsActive()
        {
            return active.Value;
        }

        virtual public List<CorrectionAction> CreateCorrections()
        {

            List<CorrectionAction> actions = new List<CorrectionAction>();
            actions.Add(new CorrectionModeAction(CorrectionType, IsActive()));

            return actions;
        }

        /** calls all executeAction */
        virtual public bool ExecuteCorrection(CorrectionAction action)
        {
            if (action is CorrectionModeAction)
            {
                return Execute((CorrectionModeAction)action);
            }
            else
            {
                log.Debug("This correction action is not registered.");
                return false;
            }
        }

        private bool Execute(CorrectionModeAction action)
        {


            if (!IsActive())
            {
                string text = LocalText.GetText("CorrectionModeActivate",
                        GetRoot.PlayerManager.CurrentPlayer.Id,
                        LocalText.GetText(CorrectionType.Name)
                );
                ReportBuffer.Add(this, text);
                DisplayBuffer.Add(this, text);
            }
            else
            {
                ReportBuffer.Add(this, LocalText.GetText("CorrectionModeDeactivate",
                        GetRoot.PlayerManager.CurrentPlayer.Id,
                        LocalText.GetText(CorrectionType.Name)
                ));
            }
            active.Set(!active.Value);

            return true;
        }

        /* dummy to capture the non-supported actions */
        virtual protected bool Execute(CorrectionAction action)
        {
            log.Debug("The chosen action is not implemented in the registered manager");
            return false;
        }

        public override int GetHashCode()
        {
            return correctionType.GetHashCode();
        }

        public bool Equals(CorrectionManager cm)
        {
            return (this.Parent == cm.Parent
                    && this.correctionType == cm.correctionType);
        }
    }
}
